cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -609,34 +703,42 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
                          std::move(aggregate_schema));
     }
 
-    case substrait::Rel::RelTypeCase::kExtensionLeaf: {
-      const auto& ext = rel.extension_leaf();
-      ARROW_ASSIGN_OR_RAISE(
-          auto ext_leaf_decl,
-          conversion_options.extension_provider->MakeRel({}, ext.detail(), ext_set));
-      return ProcessEmit(ext, std::move(ext_leaf_decl), ext_leaf_decl.output_schema);
-    }
-    case substrait::Rel::RelTypeCase::kExtensionSingle: {
-      const auto& ext = rel.extension_single();
-      ARROW_ASSIGN_OR_RAISE(DeclarationInfo input,
-                            FromProto(ext.input(), ext_set, conversion_options));
-      ARROW_ASSIGN_OR_RAISE(
-          auto ext_single_decl,
-          conversion_options.extension_provider->MakeRel({input}, ext.detail(), ext_set));
-      return ProcessEmit(ext, std::move(ext_single_decl), ext_single_decl.output_schema);
-    }
+    case substrait::Rel::RelTypeCase::kExtensionLeaf:
+    case substrait::Rel::RelTypeCase::kExtensionSingle:
     case substrait::Rel::RelTypeCase::kExtensionMulti: {
-      const auto& ext = rel.extension_multi();
-      std::vector<DeclarationInfo> inputs;
-      for (const auto& input : ext.inputs()) {
-        ARROW_ASSIGN_OR_RAISE(auto input_info,
-                              FromProto(input, ext_set, conversion_options));
-        inputs.push_back(std::move(input_info));
+      ARROW_ASSIGN_OR_RAISE(auto ext_rel_info,
+                            GetExtensionRelationInfo(rel, ext_set, conversion_options));
+      const auto& ext_decl_info = ext_rel_info.decl_info;
+      auto ext_common_opt = GetExtensionRelCommon(rel);
+      bool has_emit = ext_common_opt && ext_common_opt->emit_kind_case() ==
+                                            substrait::RelCommon::EmitKindCase::kEmit;
+      if (!ext_rel_info.field_output_indices) {
+        if (!has_emit) {
+          return ProcessEmit(ext_common_opt, ext_decl_info, ext_decl_info.output_schema);
+        }
+        return Status::NotImplemented("Emit not supported by ",
```

Comment:
My initial read of this was that `ext_rel_info` should only be needed if the extension's output matched the default output expected by Substrait.  Why wouldn't we be able to handle an emit in this case?

Reply:
The default output expected by Substrait has each output column corresponding to an input column. Given the [explanation here](https://github.com/westonpace/arrow/pull/27#discussion_r1044840134), if a relation has an undefined `field_output_indices` then its output cannot match the default output expected by Substrait.

Reply:
I think my main concern is that this is the default output for an asof join.  I don't know if the default output for every extension relation will be this way.

Reply:
Actually, you can ignore my above comment, I was confusing myself.

So the problem is that a general converter logic, without any knowledge of the extension, has no way of knowing what "the default output" should be.

Could we look at the schema in the returned declaration info and assume that is the default output?

Reply:
> Could we look at the schema in the returned declaration info and assume that is the default output?

I think the problem with using this approach in general is that the output schema is not always equal to the default output expected by Substrait; indeed, this is the case with as-of-join. OTOH, just because as-of-join is a counter-example doesn't mean we cannot use this approach as the default one for most other cases. However, I think it might be too early to do so, since we have no other cases at the moment.

Reply:
> output schema is not always equal to the default output expected by Substrait

I am not sure I understand this - can you give an example? (Actually I don't understand what is "output expected by Substrait", it's confusing to me that a node can have two output schema and still trying to wrap my heads around it.)

Reply:
Exactly. Here's an example to demonstrate this. When performing an as-of-join on `TIME` by `KEY` of a left table with `TIME, KEY, VALUE1` columns and a right table with columns `TIME, KEY, VALUE2`, the resulting output schema is `TIME, KEY, VALUE1, VALUE2` whereas the default output output schema expected by Substrait is `TIME, KEY, VALUE1, TIME, KEY, VALUE2`.

Reply:
> However, I think it might be too early to do so, since we have no other cases at the moment.

This is valid and we can address this later.  So feel free to ignore the below.

My thinking is that extension authors have a choice.  They can either always output the default order, in which case they never set `field_output_indices`.  Or they can choose to sometimes or never output the default order, in which case they must set `field_output_indices`.

So the fact that the extension did not set `field_output_indices` means that the extension is asserting the output schema is the full default output for the node.

Reply:
The downside of this PR, as it is, is that the option:

> hey can either always output the default order, in which case they never set field_output_indices

is not available.  In fact, there is no reason to make `field_output_indices` optional in this case.  The extension author has no control over the plan created by the producers and so they have to assume that, sometimes, `has_emit` will be false.

Reply:
IIUC, your thinking is that the case of an undefined `field_output_indices` is equivalent to the case of a default one, i.e. where it is set to `0, 1, ..., N-1` where `N` is the total number of fields across all inputs of the relation. In this PR they are not equivalent - the former case means it is not defined at all, so if an output mapping is given in the Substrait plan then an error will be raised (see also [this post](https://github.com/westonpace/arrow/pull/27#discussion_r1044840134)). In particular, the statement should be "They can output the default order, in which case they set `field_output_indices` to `0, 1, ..., N-1`" and the other two choices are setting to a sublist of this, or to `std::nullopt` in the above former case.